text/0000-process-ext-unix.md

```diff
@@ -0,0 +1,126 @@
+calling `fork` and `exec` manually. For example, these sorts of use cases are
+not possible with the `Command` API:
+
+* The `exec` function cannot be called without `fork`. It's often useful on Unix
+  in doing this to avoid spawning processes or improve debuggability if the
+  pre-`exec` code was some form of shim.
+* Execute other flavorful functions between the fork/exec if necessary. For
+  example some proposed extensions to the standard library are [dealing with the
+  controlling tty][tty] or dealing with [session leaders][session]. In theory
+  any sort of arbitrary code can be run between these two syscalls, and it may
+  not always be the case the standard library can provide a suitable
+  abstraction.
+
+[tty]: https://github.com/rust-lang/rust/pull/28982
+[session]: https://github.com/rust-lang/rust/pull/26470
+
+Note that neither of these pieces of functionality are possible on Windows as
+there is no equivalent of the `fork` or `exec` syscalls in the standard APIs, so
+these are specifically proposed as methods on the Unix extension trait.
+
+# Detailed design
+[design]: #detailed-design
+
+The following two methods will be added to the
+`std::os::unix::process::CommandExt` trait:
+
+```rust
+/// Performs all the required setup by this `Command`, followed by calling the
+/// `execvp` syscall.
+///
+/// On success this function will not return, and otherwise it will return an
+/// error indicating why the exec (or another part of the setup of the
+/// `Command`) failed.
+///
+/// Note that the process may be in a "broken state" if this function returns in
+/// error. For example the working directory, environment variables, signal
+/// handling settings, various user/group information, or aspects of stdio
+/// file descriptors may have changed. If a "transactional spawn" is required to
+/// gracefully handle errors it is recommended to use the cross-platform `spawn`
+/// instead.
+fn exec(&mut self) -> io::Error;
+
+/// Schedules a closure to be run just before the `exec` function is invoked.
+///
+/// This closure will be run in the context of the child process after the
+/// `fork` and other aspects such as the stdio file descriptors and working
+/// directory have successfully been changed. Note that this is often a very
+/// constrained environment where normal operations like `malloc` or acquiring a
+/// mutex are not guaranteed to work (due to other threads perhaps still running
+/// when the `fork` was run).
+///
+/// The closure is allowed to return an I/O error whose OS error code will be
+/// communicated back to the parent and returned as an error from when the spawn
+/// was requested.
+///
+/// Multiple closures can be registered and they will be called in order of
+/// their registration. If a closure returns `Err` then no further closures will
+/// be called and the spawn operation will immediately return with a failure.
+fn before_exec<F>(&mut self, f: F) -> &mut Self
```

Comment:
The `exec` and `before_exec` should be unsafe and the inline doc should warn more the user why it's important to [not use the heap](https://github.com/rust-lang/rust/blob/94aee5b7e6d1eda4c872133787ba238cbe29ecb4/src/libstd/sys/unix/process.rs#L232-L261) at all to avoid inconsistent memory layout (e.g. when malloc/free is interrupted). Cf. [reentrant closure](https://internals.rust-lang.org/t/safe-async-reentrant-closure-e-g-signal-handler/1742).

It would be nice to have a lint to check unsafe code here :)